Android EXIF reader

sklencar opened the pull request:
A new expression function introduce to read EXIF data. Currently only for Android.

Sample project: `lutraconsulting/test_exif` (should have public access on Mergin)
Supported functions: 

- `read_exif('<ABSOLUTE_PATH_TO_IMAGE>', '<EXIF_TAG_STRING>')`
-  `read_exif_img_direction('<ABSOLUTE_PATH_TO_IMAGE>')`
-  `read_exif_latitude('<ABSOLUTE_PATH_TO_IMAGE>')`
-  `read_exif_longitude('<ABSOLUTE_PATH_TO_IMAGE>')`

In case of invalid path, invalid tag or missing value, empty value is returned.
It calls java function from EXIFUtils and formats the result.

Manual test added to https://github.com/lutraconsulting/input-manual-tests/issues/5 

![Screenshot_20210323-091027](https://user-images.githubusercontent.com/6735606/112122000-2115f180-8bc0-11eb-947d-c9c43f28c657.png)

sklencar commented:
Some remarks:
- read_exif(file, tag) returns raw data as stored in the tag
- read_exif_img_direction(file) returns angle in degrees (as a number, not formatted string)
- read_exif_latitude(file), read_exif_longitude(file) return lat/lon value in degrees (again as a number, not a string)
- when calculating lat/lon from deg/min/sec please use double not float otherwise we can loose precision (float has too low precision!)
- in formatRationalNumberString() does the conversion work correctly? it looks like you are using integer division there

sklencar commented:
Remarks resolved, updated info + screenshot

saberraz commented:
Tested with the APKs and none worked for me. They all come as blank.

sklencar commented:
https://github.com/lutraconsulting/input/pull/1292 fixes applying exif expressions functions immediately after photo is captured.

https://user-images.githubusercontent.com/6735606/113024297-e0e9dc80-9186-11eb-947e-e1b28f5d846e.mp4

sklencar commented:
Rebased with master (and fix for applying default values on update)

sklencar commented:
Found one issue after rebasing with master - crash occurs after modifying an image field. Fix on a way.

sklencar commented:
Added a handling of invalid values

PeterPetrik commented on app/inputexpressionfunctions.cpp:
```diff
@@ -0,0 +1,22 @@
+
+#include "inputexpressionfunctions.h"
+
+InputExpressionFunctions::InputExpressionFunctions()
```
= default

PeterPetrik commented on app/inputexpressionfunctions.h:
```diff
@@ -0,0 +1,172 @@
+         *  - exif_tag: The officially supported EXIF TAG name.
+         * @return Formatted attribute value for given parameters, if parameters are not valid, returns empty string
+         */
+        QVariant func( const QVariantList &values, const QgsExpressionContext *, QgsExpression *, const QgsExpressionNodeFunction * ) override
+        {
+#ifdef ANDROID
+          if ( values.size() != 2 ) return QVariant();
+
+          QString filepath( values.at( 0 ).toString() );
+          QString exifTag( values.at( 1 ).toString() );
+          return AndroidUtils::getExifInfo( filepath, exifTag );
+# else
+          return QString();
+#endif
+        }
```
move impl to cpp

PeterPetrik commented on app/inputexpressionfunctions.h:
```diff
@@ -0,0 +1,172 @@
+         *  - file: Absolute path of an image that exif attribute will be read from,
+         * @return EXIF value for GPS Image Direction
+         */
+        QVariant func( const QVariantList &values, const QgsExpressionContext *, QgsExpression *, const QgsExpressionNodeFunction * ) override
+        {
+#ifdef ANDROID
+          if ( values.size() != 1 ) return QVariant();
+
+          QString filepath( values.at( 0 ).toString() );
+          QString resultString = AndroidUtils::getExifInfo( filepath, GPS_DIRECTION_TAG );
+          if ( resultString.isEmpty() )
+            return QVariant();
+
+          double resultDouble = InputUtils::convertRationalNumber( resultString );
+          // If invalid value; expecting [0-360]
+          if ( resultDouble == -1 )
+            return QVariant();
+
+          return QVariant( resultDouble );
+# else
+          return QString();
+#endif
```
move impl to cpp

PeterPetrik commented on app/inputexpressionfunctions.h:
```diff
@@ -0,0 +1,172 @@
+         *  - file: Absolute path of an image that exif attribute will be read from,
+         * @return EXIF value for GPS Latitude
+         */
+        QVariant func( const QVariantList &values, const QgsExpressionContext *, QgsExpression *, const QgsExpressionNodeFunction * ) override
+        {
+#ifdef ANDROID
+          if ( values.size() != 1 ) return QVariant();
+
+          QString filepath( values.at( 0 ).toString() );
+          QString resultString = AndroidUtils::getExifInfo( filepath, GPS_LAT_TAG );
+          if ( resultString.isEmpty() )
+            return QVariant();
+
+          return QVariant( InputUtils::convertCoordinateString( resultString ) );
+# else
+          return QString();
+#endif
+        }
```
move impl to cpp

PeterPetrik commented on app/inputexpressionfunctions.h:
```diff
@@ -0,0 +1,172 @@
+         *  - file: Absolute path of an image that exif attribute will be read from,
+         * @return EXIF value for GPS Longitude
+         */
+        QVariant func( const QVariantList &values, const QgsExpressionContext *, QgsExpression *, const QgsExpressionNodeFunction * ) override
+        {
+#ifdef ANDROID
+          if ( values.size() != 1 ) return QVariant();
+
+          QString filepath( values.at( 0 ).toString() );
+          QString resultString = AndroidUtils::getExifInfo( filepath, GPS_LON_TAG );
+          if ( resultString.isEmpty() )
+            return QVariant();
+
+          return QVariant( InputUtils::convertCoordinateString( resultString ) );
+# else
+          return QString();
+#endif
+        }
```
move impl to cpp

PeterPetrik commented on app/inputexpressionfunctions.h:
```diff
@@ -0,0 +1,172 @@
+        QgsScopedExpressionFunction *clone() const override { return new ReadExifLongitude();}
+    };
+
+    InputExpressionFunctions();
```
why we need custom constructor, if needed, move to the top of the file also with the register Functions function

PeterPetrik commented on app/inputexpressionfunctions.h:
```diff
@@ -0,0 +1,172 @@
+const static QString GPS_LON_TAG = "GPSLongitude";
+const static QString GPS_LAT_TAG = "GPSLatitude";
+
+class InputExpressionFunctions
```
add doc

PeterPetrik commented on app/inpututils.cpp:
```diff
@@ -140,6 +140,31 @@ void InputUtils::setExtentToFeature( const QgsQuickFeatureLayerPair &pair, QgsQu
   mapSettings->setExtent( currentExtent );
 }
 
+double InputUtils::convertCoordinateString( const QString &rationalValue )
```
create unit test

PeterPetrik commented on app/inpututils.cpp:
```diff
@@ -140,6 +140,31 @@ void InputUtils::setExtentToFeature( const QgsQuickFeatureLayerPair &pair, QgsQu
+  return result;
+}
+
+double InputUtils::convertRationalNumber( const QString &rationalValue )
```
create unit test

PeterPetrik commented on app/inpututils.cpp:
```diff
@@ -140,6 +140,31 @@ void InputUtils::setExtentToFeature( const QgsQuickFeatureLayerPair &pair, QgsQu
+
+  double numerator = number.at( 0 ).toDouble();
+  double denominator = number.at( 1 ).toDouble();
+  return numerator / denominator;
```
whatif denominator == 0

PeterPetrik commented on app/inputexpressionfunctions.h:
```diff
@@ -0,0 +1,172 @@
+
+    InputExpressionFunctions();
+
+    static void registerFunctions();
```
I am not convinced we need another class `InputExpressionFunctions()`. Cannot be `registerFunctions` just another static function of InputUtils ?

sklencar commented:
Unit tests were omitted in this PR for now (and created a generic ticket for implementing tests for utils functions).
However, added more checks for util functions used with EXIF functions.

PeterPetrik commented:
test: https://github.com/lutraconsulting/input-manual-tests/issues/5 -> Test G
docs: TODO

sklencar commented:
Added docs: https://github.com/lutraconsulting/input-docs/pull/29
Added manual test: https://github.com/lutraconsulting/input-manual-tests/issues/5 (Test H)
